lightning/src/ln/onion_route_tests.rs

```diff
@@ -577,6 +577,170 @@ fn test_onion_failure() {
+	assert!(hops[1].node_features.supports_variable_length_onion());
+	// Assert that the hop between `nodes[2]` and `nodes[3]` defaults to supporting variable length
+	// onions, as `nodes[0]` has no `NodeAnnouncementInfo` `features` for `node[3]`, and no `InvoiceFeatures`
+	// for the `payment_params`, which would otherwise have been used.
+	assert!(hops[2].node_features.supports_variable_length_onion());
+	// Note that we do not asset that `hops[0]` (the channel between `nodes[0]` and `nodes[1]`)
+	// supports variable length onions, as the `InitFeatures` exchanged in the init message when
+	// between the nodes will be used when creating the route. We therefore do not default to
+	// supporting variable length onions for that hop, as the `InitFeatures` in this case are
+	// `InitFeatures::known()`.
+
+	let unannounced_chan = &nodes[4].node.list_usable_channels()[0];
+
+	let last_hop = RouteHint(vec![RouteHintHop {
+		src_node_id: nodes[3].node.get_our_node_id(),
+		short_channel_id: unannounced_chan.short_channel_id.unwrap(),
+		fees: RoutingFees {
+			base_msat: 0,
+			proportional_millionths: 0,
+		},
+		cltv_expiry_delta: 42,
+		htlc_minimum_msat: None,
+		htlc_maximum_msat: None,
+	}]);
+
+	let unannounced_chan_params = PaymentParameters::from_node_id(nodes[4].node.get_our_node_id()).with_route_hints(vec![last_hop]);
+	let (unannounced_route, _, _, _) = get_route_and_payment_hash!(
+		origin_node, nodes[4], unannounced_chan_params, 10_000, TEST_FINAL_CLTV);
+
+	let unannounced_chan_hop = &unannounced_route.paths[0][3];
+	// Ensure that `nodes[4]` doesn't exist in `nodes[0]`'s `network_graph`, as it's not public.
+	assert!(&network_graph.read_only().nodes().get(&NodeId::from_pubkey(&nodes[4].node.get_our_node_id())).is_none());
+	// Assert that the hop between `nodes[3]` and `nodes[4]` defaults to supporting variable length
+	// onions, even though `nodes[4]` as `nodes[0]` doesn't exists in `nodes[0]`'s `network_graph`,
+	// and no `InvoiceFeatures` for the `payment_params` exists, which would otherwise have been
+	// used.
+	assert!(unannounced_chan_hop.node_features.supports_variable_length_onion());
+
+	let cur_height = nodes[0].best_block_info().1 + 1;
+	let (announced_route_payloads, _htlc_msat, _htlc_cltv) = onion_utils::build_onion_payloads(&announced_route.paths[0], 40000, &None, cur_height, &None).unwrap();
+	let (unannounced_route_paylods, _htlc_msat, _htlc_cltv) = onion_utils::build_onion_payloads(&unannounced_route.paths[0], 40000, &None, cur_height, &None).unwrap();
+
+	for onion_payloads in vec![announced_route_payloads, unannounced_route_paylods] {
+		for onion_payload in onion_payloads.iter() {
+			match onion_payload.format {
+				msgs::OnionHopDataFormat::Legacy {..} => {
+					panic!("Generated a `msgs::OnionHopDataFormat::Legacy` payload, even though that shouldn't have happend.");
+				}
+				_ => {}
+			}
+		}
+	}
+}
+
+#[test]
+fn test_do_not_use_default_to_tlv_onions_if_other_features_not_supporting_them_exists() {
```

Comment:
To name somewhat similar to the previous test, how about: `test_do_not_default_to_onion_payload_tlv_format_when_unsupported`